[GTK] Deprecate event parameter of WebKitWebView::context-menu signal

carlosgcampos opened the pull request:
#### 07a9bc7be85d6549f876018d1886a8b4335acc25
<pre>
[GTK] Deprecate event parameter of WebKitWebView::context-menu signal
<a href="https://bugs.webkit.org/show_bug.cgi?id=248078">https://bugs.webkit.org/show_bug.cgi?id=248078</a>

Reviewed by Michael Catanzaro.

Deprecate it in current API and remove it for GTK4 API. Add
webkit_context_menu_get_event() instead.

* Source/WebKit/Shared/API/glib/WebKitContextMenu.cpp:
(webkitContextMenuSetEvent):
(webkit_context_menu_get_event):
* Source/WebKit/Shared/API/glib/WebKitContextMenuPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitContextMenu.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkit_web_view_class_init):
(webkitWebViewPopulateContextMenu):
(gValueGetEvent): Deleted.
(webkitWebViewContextMenuMarshal): Deleted.
(webkitWebViewContextMenuMarshalVa): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitGLib/WebExtensionTest.cpp:
(contextMenuCallback):
* Tools/TestWebKitAPI/Tests/WebKitGtk/TestContextMenu.cpp:
(beforeAll):

Canonical link: <a href="https://commits.webkit.org/256886@main">https://commits.webkit.org/256886@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/201adb26455b9815cc9dabe98087b7956e3f2ab2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97122 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6388 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30245 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106639 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/166911 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101093 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6642 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35122 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89511 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/103329 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102790 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4985 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83741 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/31981 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86834 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/74870 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/410 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/394 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/21593 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5192 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44097 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2327 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1634 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/40908 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/Shared/API/glib/WebKitContextMenu.cpp:
```diff
@@ -372,3 +390,37 @@ GVariant* webkit_context_menu_get_user_data(WebKitContextMenu* menu)
+ *
+ * Gets the #GdkEvent that triggered the context menu. This function only returns a valid
+ * #GdkEvent when called for a #WebKitContextMenu passed to #WebKitWebView::context-menu
+ * signal, in all other cases %NULL is returned.
```
```suggestion
 * signal; in all other cases, %NULL is returned.
```

mcatanzaro commented on Source/WebKit/Shared/API/glib/WebKitContextMenuPrivate.h:
```diff
@@ -22,8 +22,23 @@
+#else
+#include <WebCore/GUniquePtrGtk.h>
+#endif
+#endif
```
I'd prefer to write:

```suggestion
#if PLATFORM(GTK)
#include <WebCore/GRefPtrGtk.h>
#include <WebCore/GUniquePtrGtk.h>
#endif
```

When you are excessive with conditionalizing the includes, it becomes more likely that we'll accidentally break the build in the future.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -2076,11 +2011,12 @@ static void webkit_web_view_class_init(WebKitWebViewClass* webViewClass)
         G_TYPE_BOOLEAN, 1, /* number of parameters */
         WEBKIT_TYPE_FILE_CHOOSER_REQUEST);
 
+    // FIXME: We need a solution to split the documentation that doesn't break introspection.
```
Indeed. Using #if within the doc comments will not work. Does it work if you duplicate the entire comment?

```
#if USE(GTK4)
/**
 * ...
 */
#else
/**
 * ...
 */
#endif
```

My guess is probably not, but it's worth trying, because that keeps the documentation next to the code, which is the only way we have a chance to keep it from becoming stale.

If that fails, then we've no choice but to split this doc comment into separate files and pass different sets of files to the introspection scanner, one WebKitWebViewGtk3.cpp for GTK 3, one for GTK 4, and one for WPE. I think it's OK to do this later rather than now, as long as you report a bug that blocks https://bugs.webkit.org/show_bug.cgi?id=GTK4 so we don't forget about this FIXME.

carlosgcampos replied:
We need to split in different files, that's why I added the comment instead of doing it as part of this commit.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -2111,21 +2047,8 @@ static void webkit_web_view_class_init(WebKitWebViewClass* webViewClass)
-     * signal was used to show the context menu.
-     * </para></listitem>
-     * </itemizedlist>
+     * The @event parameter is now deprecated, use webkit_context_menu_get_event() to get the
```
```suggestion
     * The @event parameter is now deprecated. Use webkit_context_menu_get_event() to get the
```

Or:

```suggestion
     * The @event parameter is now deprecated; use webkit_context_menu_get_event() to get the
```

mcatanzaro reviewed (approved):
Good.

We'll need to fix the option menus too, https://bugs.webkit.org/show_bug.cgi?id=246716, but that can be something for later.

mcatanzaro commented on Tools/TestWebKitAPI/Tests/WebKitGtk/TestContextMenu.cpp:
```diff
@@ -66,6 +66,7 @@ class ContextMenuTest: public WebViewTest {
     {
         g_assert_true(WEBKIT_IS_CONTEXT_MENU(contextMenu));
         test->assertObjectIsDeletedWhenTestFinishes(G_OBJECT(contextMenu));
+        g_assert_true(webkit_context_menu_get_event(contextMenu) == event);
```
```suggestion
#if PLATFORM(GTK)
        g_assert_true(webkit_context_menu_get_event(contextMenu) == event);
#endif
```

carlosgcampos replied:
This shouldn't be needed because this test is GTK specific

mcatanzaro commented on Tools/TestWebKitAPI/Tests/WebKitGLib/WebExtensionTest.cpp:
```diff
@@ -272,6 +272,8 @@ static GVariant* serializeNode(JSCValue* node)
 
 static gboolean contextMenuCallback(WebKitWebPage* page, WebKitContextMenu* menu, WebKitWebHitTestResult* hitTestResult, gpointer)
 {
+    g_assert_null(webkit_context_menu_get_event(menu));
```
```suggestion
#if PLATFORM(GTK)
    g_assert_null(webkit_context_menu_get_event(menu));
#endif
```

mcatanzaro commented on Tools/TestWebKitAPI/Tests/WebKitGtk/TestContextMenu.cpp:
```diff
@@ -326,7 +327,7 @@ class ContextMenuDefaultTest: public ContextMenuTest {
     {
     }
 
-    bool contextMenu(WebKitContextMenu* contextMenu, GdkEvent* event, WebKitHitTestResult* hitTestResult)
+    bool contextMenu(WebKitContextMenu* contextMenu, GdkEvent*, WebKitHitTestResult* hitTestResult)
```
Careful, this only works because we don't have bots running the API tests with the modern API. You have to do:

```
#if USE(GTK4)
bool contextMenu(WebKitContextMenu* contextMenu, WebKitHitTestResult* hitTestResult)
#else
bool contextMenu(WebKitContextMenu* contextMenu, GdkEvent*, WebKitHitTestResult* hitTestResult)
#endif
```

carlosgcampos replied:
Right!

carlosgcampos replied:
Actually, we have all tests in this file running only for gtk3. I think we need a new bug to enable these tests for gtk4.

carlosgcampos committed:
[GTK] Deprecate event parameter of WebKitWebView::context-menu signal
https://bugs.webkit.org/show_bug.cgi?id=248078

Reviewed by Michael Catanzaro.

Deprecate it in current API and remove it for GTK4 API. Add
webkit_context_menu_get_event() instead.

* Source/WebKit/Shared/API/glib/WebKitContextMenu.cpp:
(webkitContextMenuSetEvent):
(webkit_context_menu_get_event):
* Source/WebKit/Shared/API/glib/WebKitContextMenuPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitContextMenu.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkit_web_view_class_init):
(webkitWebViewPopulateContextMenu):
(gValueGetEvent): Deleted.
(webkitWebViewContextMenuMarshal): Deleted.
(webkitWebViewContextMenuMarshalVa): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitGLib/WebExtensionTest.cpp:
(contextMenuCallback):
* Tools/TestWebKitAPI/Tests/WebKitGtk/TestContextMenu.cpp:
(beforeAll):

Canonical link: https://commits.webkit.org/256886@main
